apps/vetrina/src/Vetrina/Main.purs

```diff
@@ -268,152 +240,64 @@ vetrinaContainer children =
-        when (userHasPackage product.packageName $ map _.package user.subs)
-          $ ExceptT $ pure $ Left SubscriptionExists
-        order      <- ExceptT $ createOrder user product
-        paymentUrl <- ExceptT $ payOrder order self.state.form.paymentMethod
-        pure { paymentUrl, order, user }
-      case eitherRes of
-        Right { paymentUrl, order, user } ->
-          liftEffect do
-            let newState = self.state { purchaseState = CapturePayment paymentUrl
-                                      , newOrder      = Just order
-                                      , user          = Just user
-                                      , orderFailure  = Nothing
-                                      }
-            self.setState \_ -> newState
-            -- NOTE: We need to pass the updated state here, not `self.state`.
-            startOrderPoller self.setState newState order
-        Left err
-          | UnrecognizedError e <- err ->
-            liftEffect do
-              logger.error $ Error.orderError $ "Failed to place an order: " <> e
-              self.setState _ { purchaseState = PurchaseFailed }
-          | EmailInUse <- err -> liftEffect $ self.setState _ { accountStatus = ExistingAccount, orderFailure = Just EmailInUse }
-          | SubscriptionExists <- err -> liftEffect $ self.setState _ { purchaseState = PurchaseSubscriptionExists }
-          | otherwise -> liftEffect $ self.setState _ { orderFailure = Just err }
-  )
+-- TODO: Validate `acceptLegalTerms` of `NewAccountForm`
+mkPurchaseWithNewAccount :: Self -> NewPurchase.NewAccountForm -> Effect Unit
+mkPurchaseWithNewAccount self validForm = mkPurchase self validForm $ createNewAccount self validForm.emailAddress
+
+mkPurchaseWithExistingAccount :: Self -> NewPurchase.ExistingAccountForm -> Effect Unit
+mkPurchaseWithExistingAccount self validForm =
+  mkPurchase self validForm $ loginToExistingAccount self validForm.emailAddress validForm.password
+
+mkPurchaseWithLoggedInAccount :: Self -> User.User -> { | NewPurchase.PurchaseParameters } -> Effect Unit
+mkPurchaseWithLoggedInAccount self user validForm = mkPurchase self validForm $ (pure $ Right user)
+
+mkPurchase
+  :: forall r
+  . Self
+  -> { productSelection :: Maybe Product, paymentMethod :: Maybe PaymentMethod | r }
+  -> Aff (Either OrderFailure User.User)
+  -> Effect Unit
+mkPurchase self@{ state: { logger } } validForm affUser = Aff.launchAff_ $ Spinner.withSpinner (self.setState <<< Spinner.setSpinner) do
+  eitherRes <- runExceptT do
+    user <- ExceptT affUser
+    ExceptT $ Right unit <$ (liftEffect $ logger.setUser $ Just user)
+
+    product       <- ExceptT $ pure $ note (FormFieldError [ ProductSelection ]) validForm.productSelection
+    paymentMethod <- ExceptT $ pure $ note (FormFieldError [ PaymentMethod ])    validForm.paymentMethod
+
+    when (userHasPackage product.packageName $ map _.package user.subs)
+      $ ExceptT $ pure $ Left SubscriptionExists
```

Comment:
Here a lot of things are packaged just to fit into an `ExceptT` object.
I'm wondering isn't there a better way of writing this code?

Reply:
Does @zudov have any thoughts?